state/cleanup.go

```diff
@@ -292,28 +292,43 @@ func (st *State) cleanupStorageForDyingModel(cleanupArgs []bson.Raw) (err error)
 		return errors.Trace(err)
 	}
 	destroyStorage := sb.DestroyStorageInstance
+	destroyStorageFromArg := func() error {
+		var destroyStorageFlag bool
+		if err := cleanupArgs[0].Unmarshal(&destroyStorageFlag); err != nil {
+			return errors.Annotate(err, "unmarshalling cleanup arg 'destroyStorage'")
+		}
+		if !destroyStorageFlag {
+			destroyStorage = sb.ReleaseStorageInstance
+		}
+		return nil
+	}
+
+	var force bool
 	switch n := len(cleanupArgs); n {
 	case 0:
 		// Old cleanups have no args, so follow the old
 		// behaviour: destroy the storage.
 	case 1:
-		var destroyStorageFlag bool
-		if err := cleanupArgs[0].Unmarshal(&destroyStorageFlag); err != nil {
-			return errors.Annotate(err, "unmarshalling cleanup args")
+		if err := destroyStorageFromArg(); err != nil {
+			return err
 		}
-		if !destroyStorageFlag {
-			destroyStorage = sb.ReleaseStorageInstance
+	case 2:
+		if err := destroyStorageFromArg(); err != nil {
```

Comment:
You could avoid duplicating the `destroyStorageFromArg` call by having case 2 first and then using `fallthrough` to process `arg[1]`

```go
case 2:
        if err := cleanupArgs[1].Unmarshal(&force); err != nil {
			return errors.Annotate(err, "unmarshalling cleanup arg 'force'")
        }
        fallthrough
case 1:
        if err := destroyStorageFromArg(); err != nil {
                return err
        }
}
```

Reply:
On second thought, that seems a bit gross. It would be better to just use multiple if statements:

```go
if len(cleanupArgs) >= 1 {
        if err := destroyStorageFromArg(); err != nil {
                return err
        }
}
if len(cleanupArgs) >= 2 {
        if err := cleanupArgs[1].Unmarshal(&force); err != nil {
			return errors.Annotate(err, "unmarshalling cleanup arg 'force'")
        }
}
```

Reply:
true :) agreed! I like the 2nd version better - it's clean. Changed where appropriate and applicable.